### PR TITLE
GTK# / Xamarin.Forms.Gtk implementations (reloaded)

### PR DESCRIPTION
**Pull request by jeremyVignelles**
replaces #2 
Due to the recent changes in the libvlcsharp, too many conflicts prevented the proper merge of #2.

This is a new, better implementation. It should fix cross-platform build issues, but it still needs to be confirmed. (My msbuild on my linux has been broken by an update : https://github.com/mono/linux-packaging-msbuild/issues/8 )

Forms.GTK# also works! The initialization bug does not affect us anymore due to the fact that the media player is now given externally.

Checklist

- [X] Write the code for the Gtk# implementation
- [X] Write the code for Xamarin.Forms using the GTK# backend
- [X] Make it work on Windows
- [x] Make it work on Linux
- [ ] Make it work on MacOS
- [ ] Check that the Forms.Gtk sample works on all three platforms
- [ ] Write installation instructions on all 3 platforms

/cc @mfkl @Ch4rg3r

### CODE REVIEW

File: LibVLCSharp.Gtk/VideoView.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+using LibVLCSharp.Shared;
+using Object = Gtk.Object;
+
+namespace LibVLCSharp.Gtk
+{
+    public class VideoView : DrawingArea, IVideoView
+    {
+        private MediaPlayer _mediaPlayer;
+
+        public VideoView(MediaPlayer mediaPlayer = null)
```
Comment (mfkl) [inline LibVLCSharp.Gtk/VideoView.cs]: The `VideoView` ctor on the other platforms does not have this optional parameters. Either we remove it here or we add it in the other `VideoView` implementations.

File: LibVLCSharp.Gtk/VideoView.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        }
+
+        void Attach()
+        {
+            if (!this.IsRealized || this._mediaPlayer == null)
+            {
+                return;
+            }
+
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Comment (mfkl) [inline LibVLCSharp.Gtk/VideoView.cs]: Do we need to perform older checks for net40 targets here?

**Comment by mfkl** [comment]
https://github.com/videolan/libvlcsharp/blob/1c0655b9be3807c29bd7a5bd45d9c15cef05ae4b/LibVLCSharp/Shared/Core.cs#L157 returns 4 (Unix) on macOS. Only way around it is to start a new process and check for the name "darwin", ugh

**Comment by mfkl** [comment]
This fails on GTK.Mac
https://github.com/videolan/libvlcsharp/blob/1c0655b9be3807c29bd7a5bd45d9c15cef05ae4b/LibVLCSharp/Shared/Core.cs#L132

No idea why, no logs... same libvlc gets loaded fine in the Xamarin.Mac Cocoa sample, and I got it working before... meh

File: LibVLCSharp.Gtk/VideoView.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
+﻿using System;
+using System.Runtime.InteropServices;
+using Gdk;
+using Gtk;
+using LibVLCSharp.Shared;
+using Object = Gtk.Object;
```
Comment (jeremyVignelles) [inline LibVLCSharp.Gtk/VideoView.cs]: WTF is that? x) need to remove that...

File: LibVLCSharp.Gtk/VideoView.cs (hunk 1 above)
Comment (jeremyVignelles) [reply LibVLCSharp.Gtk/VideoView.cs]: I found it easier to use in the Gtk# sample. If you want to use that in the WPF control, it's hard to pass a constructor parameter.

I'm ok with both ways, just tell me

**Comment by mfkl** [comment]
>  == 6 seems magic, while there is an enum value (see what I did on linux)

It was needed magic, unfortunately it does not work. https://github.com/dotnet/corefx/issues/19694
NET Classic minimum target on Mac will be net471

**Comment by mfkl** [comment]
- libvlc for mac seems to be included 1/2 times when rebuilding... weird. Always fine with netcore targets
- I managed to build and run the sample on mac, only from CLI. Running with VS is not possible as the dylib is not copied to output (but building with CLI it is, _most of the time_)
- dlopen of libvlc on Linux is missing.
- And last but not least, here's a weird one: The second I touch the GTK videoview on mac with the mouse cursor, it crashes with

```
Stack overflow in unmanaged: IP: 0x107261f69, fault addr: 0x7ffeec73eff8
Stack overflow: IP: 0x107261f69, fault addr: 0x7ffeec735ff8
Stacktrace:
  at <unknown> <0xffffffff>
  at Gtk.Application.Run () [0x00001] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/gtk-sharp-None/gtk/Application.cs:145
  <...>
  at LibVLCSharp.GTK.Sample.Program.Main () [0x0009d] in /Users/Martz/Projects/LibVLCSharp/Samples/LibVLCSharp.GTK.Sample/Program.cs:35
  at (wrapper runtime-invoke) object.runtime_invoke_void (object,intptr,intptr,intptr) [0x0004d] in <71f4e3d7d65e42b0a699d02e59766d3c>:0
```

File: LibVLCSharp.Gtk/VideoView.cs (hunk 2 above)
Comment (jeremyVignelles) [reply LibVLCSharp.Gtk/VideoView.cs]: Xamarin.Forms.Platform.GTK is compatible with 4.5. Shall we lower the version?

**Comment by jeremyVignelles** [comment]
Merged @mfkl 's suggestions into my branch and rebased / squashed the commits.

@mfkl : What's left to do on your side with Mac?

For linux, everything builds fine thanks to instructions here : https://github.com/mono/linux-packaging-msbuild/pull/7, but I still can't build the Forms.Gtk.Sample : 

```
Microsoft (R) Build Engine version 15.8.0.0 ( Thu Nov 15 18:28:44 CET 2018) pour Mono
Copyright (C) Microsoft Corporation. Tous droits réservés.

La génération a démarré 15/11/2018 21:29:31.
Projet "/data/projects/dotnet/libvlcsharp/Samples/Forms/LibVLCSharp.Forms.Sample.GTK/LibVLCSharp.Forms.Sample.GTK.csproj" sur le nœud 1 (cibles par défaut).
/data/projects/dotnet/libvlcsharp/Samples/Forms/LibVLCSharp.Forms.Sample.GTK/LibVLCSharp.Forms.Sample.GTK.csproj(151,5): error : Ce projet fait référence à des packages NuGet qui sont manquants sur cet ordinateur. Utilisez l'option de restauration des packages NuGet pour les télécharger. Pour plus d'informations, consultez http://go.microsoft.com/fwlink/?LinkID=322105. Le fichier manquant est : ../../../packages/Xamarin.Forms.3.2.0.871581/build/netstandard2.0/Xamarin.Forms.props.
Génération du projet "/data/projects/dotnet/libvlcsharp/Samples/Forms/LibVLCSharp.Forms.Sample.GTK/LibVLCSharp.Forms.Sample.GTK.csproj" terminée (cibles par défaut) -- ÉCHEC.

ÉCHEC de la build.

"/data/projects/dotnet/libvlcsharp/Samples/Forms/LibVLCSharp.Forms.Sample.GTK/LibVLCSharp.Forms.Sample.GTK.csproj" (cible par défaut) (1) ->
(EnsureNuGetPackageBuildImports cible) -> 
  /data/projects/dotnet/libvlcsharp/Samples/Forms/LibVLCSharp.Forms.Sample.GTK/LibVLCSharp.Forms.Sample.GTK.csproj(151,5): error : Ce projet fait référence à des packages NuGet qui sont manquants sur cet ordinateur. Utilisez l'option de restauration des packages NuGet pour les télécharger. Pour plus d'informations, consultez http://go.microsoft.com/fwlink/?LinkID=322105. Le fichier manquant est : ../../../packages/Xamarin.Forms.3.2.0.871581/build/netstandard2.0/Xamarin.Forms.props.

    0 Avertissement(s)
    1 Erreur(s)

Temps écoulé 00:00:00.45
```

It tells me it can't find `../../../packages/Xamarin.Forms.3.2.0.871581/build/netstandard2.0/Xamarin.Forms.props`, despite having run `msbuild /t:Restore`, which does nothing (or at least doesn't install anything at the path specified). Any idea @jsuarezruiz ?

**Comment by mfkl** [comment]
> @mfkl : What's left to do on your side with Mac?

The stackoverflow when the mouse cursor touches the GTK `DrawingArea` would be the last blocker. The rest is just annoying but we can ship without (CLI only, VS4Mac crashes when editing GTK projects).
I have tried a few things on the libvlc side (i.e. `mp.EnableMouseInput` false/true) and on the GTK side but no clue... Have to use gdb. 
Note that I can grab and move the window no problem, it crashes only when the cursor comes in contact with the `DrawingArea`.

For linux, can you write a small instruction guide to get proper libvlc (for ubuntu at least) and whatever else is needed to get the sample running? Will add it to the gitlab wiki page.

File: Samples/Forms/LibVLCSharp.Forms.Sample.GTK/LibVLCSharp.Forms.Sample.GTK.csproj [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    </ProjectReference>
+    <ProjectReference Include="..\LibVLCSharp.Forms.Sample\LibVLCSharp.Forms.Sample.csproj">
+      <Project>{a7f1b21f-ca82-4fc7-871f-2be5b9d6da70}</Project>
+      <Name>LibVLCSharp.Forms.Sample</Name>
+    </ProjectReference>
+  </ItemGroup>
+  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <Target Name="EnsureNuGetPackageBuildImports" BeforeTargets="PrepareForBuild">
+    <PropertyGroup>
+      <ErrorText>Ce projet fait référence à des packages NuGet qui sont manquants sur cet ordinateur. Utilisez l'option de restauration des packages NuGet pour les télécharger. Pour plus d'informations, consultez http://go.microsoft.com/fwlink/?LinkID=322105. Le fichier manquant est : {0}.</ErrorText>
```
[the remark this answers is not in the document]
Comment (jeremyVignelles) [reply Samples/Forms/LibVLCSharp.Forms.Sample.GTK/LibVLCSharp.Forms.Sample.GTK.csproj]: What's the english version of those messages?

File: LibVLCSharp.Forms.Platforms.GTK/LibVLCSharp.Forms.Platforms.GTK.csproj [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  <ItemGroup>
+    <PackageReference Include="Xamarin.Forms.Platform.GTK" Version="3.2.0.871581" />
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\LibVLCSharp.Forms\LibVLCSharp.Forms.csproj" />
+    <ProjectReference Include="..\LibVLCSharp.GTK\LibVLCSharp.GTK.csproj" />
+  </ItemGroup>
+  <ItemGroup>
+    <Reference Include="gdk-sharp">
+      <SpecificVersion>False</SpecificVersion>
```
Comment (mfkl) [inline LibVLCSharp.Forms.Platforms.GTK/LibVLCSharp.Forms.Platforms.GTK.csproj]: `<SpecificVersion>False</SpecificVersion>` this seem to be removed systematically by VS4Mac FYI.

File: Samples/Forms/LibVLCSharp.Forms.Sample.GTK/LibVLCSharp.Forms.Sample.GTK.csproj (hunk 4 above)
[the remark this answers is not in the document]
Comment (mfkl) [reply Samples/Forms/LibVLCSharp.Forms.Sample.GTK/LibVLCSharp.Forms.Sample.GTK.csproj]: https://github.com/videolan/libvlcsharp/blob/9cf2464b623457930513a2799e1c3d4a9a6292b3/Samples/LibVLCSharp.Android.Sample/LibVLCSharp.Android.Sample.csproj#L102

I see most samples have it in french as well

**Comment by jeremyVignelles** [comment]
I managed to make the Xamarin.Forms sample work on linux: I fixed the project reference path (linux is case-sensitive) and restored the packages with `nuget install -OutputDirectory ../../../packages` in the Samples/Forms/LibVLCSharp.Forms.Sample.GTK folder.

It works well, but it's quite manual. Shall we switch to a SDK-style project instead? Shall we make a template project?

**Comment by jeremyVignelles** [comment]
> Sure. Are you saying that would help fix the manual restore part?
I guess: instead of `nuget install`, `dotnet build` automatically restores the packages. I need to test if it works

> Not sure what you mean. Sample project?
A repo project that can be used to `dotnet new VideoLAN.LibVLCSharp.Forms.GTK`. I need to look how it works.

File: Samples/Forms/LibVLCSharp.Forms.Sample.GTK/LibVLCSharp.Forms.Sample.GTK.csproj (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeremyVignelles) [reply Samples/Forms/LibVLCSharp.Forms.Sample.GTK/LibVLCSharp.Forms.Sample.GTK.csproj]: This project is now SDK-Style, so no problem anymore.

File: LibVLCSharp.Gtk/VideoView.cs (hunk 2 above)
Comment (jeremyVignelles) [reply LibVLCSharp.Gtk/VideoView.cs]: Let's use 4.7.1 for now, and see if someone is interested.

**Comment by jeremyVignelles** [comment]
Current status: Almost done.

Remaining:
- GTK# stackoverflows on mac due to https://github.com/mono/gtk-sharp/issues/257 , will probably release with the bug...
- Docs
- More docs
- Squash
- Merge :tada:

**Comment by jbkempf** [comment]
I'd argue you should merge without macOS for now.

**Comment by mfkl** [comment]
Merged https://github.com/videolan/libvlcsharp/commit/a927af513addeacb7c17541d06c6b7d16bfef9c0.

Good job 👍

**Pull request closed without merging**